File: legacy_backend/core/services/ai_service.py
"""
AI service for content processing using OpenAI-compatible APIs.

Provides translation, summarization, and custom prompt processing
with structured JSON output and retry logic.
"""

import json
import logging
import time
from datetime import timedelta
from typing import Any

import requests
from django.conf import settings
from django.utils import timezone

logger = logging.getLogger(__name__)


class AIServiceError(Exception):
    """Base exception for AI service errors."""

    pass


class AIQuotaExceededError(AIServiceError):
    """Raised when user exceeds AI quota."""

    pass


class AIService:
    """
    OpenAI-compatible AI service for content processing.

    Uses structured output (JSON mode) to ensure reliable parsing.
    Implements retry logic with exponential backoff.
    """

    def __init__(self):
        self.api_url = settings.OPENAI_API_URL.rstrip("/")
        self.api_key = settings.OPENAI_API_KEY
        self.model = settings.AI_MODEL
        self.temperature = settings.AI_TEMPERATURE
        self.max_tokens = settings.AI_MAX_TOKENS
        self.timeout = settings.AI_REQUEST_TIMEOUT
        self.max_retries = settings.AI_MAX_RETRIES
        self.retry_delay = settings.AI_RETRY_DELAY

        if not self.api_key:
            raise AIServiceError("OPENAI_API_KEY not configured")

    def _make_request(
        self,
        messages: list[dict[str, str]],
        response_format: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Make API request with retry logic.

        Args:
            messages: List of message dicts with 'role' and 'content'
            response_format: Optional response format schema

        Returns:
            Parsed JSON response

        Raises:
            AIServiceError: On API error or max retries exceeded
        """
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

        payload = {
            "model": self.model,
            "messages": messages,
            "temperature": self.temperature,
            "max_tokens": self.max_tokens,
        }

        # Add response format if provided (for structured output)
        if response_format:
            payload["response_format"] = response_format

        last_error = None

        for attempt in range(self.max_retries):
            try:
                response = requests.post(
                    f"{self.api_url}/chat/completions",
                    headers=headers,
                    json=payload,
                    timeout=self.timeout,
                )
                response.raise_for_status()

                data = response.json()
                content = data["choices"][0]["message"]["content"]

                # Parse JSON response
                if response_format:
                    return json.loads(content)
                return {"content": content}

            except requests.exceptions.RequestException as e:
                last_error = e
                logger.warning(
                    f"AI request failed (attempt {attempt + 1}/{self.max_retries}): {e}"
                )

                if attempt < self.max_retries - 1:
                    # Exponential backoff
                    delay = self.retry_delay * (2**attempt)
                    time.sleep(delay)
                    continue
                break

            except (KeyError, json.JSONDecodeError) as e:
                last_error = e
                logger.error(f"Failed to parse AI response: {e}")
                break

        raise AIServiceError(
            f"AI request failed after {self.max_retries} retries: {last_error}"
        )

    def translate(
        self, content: str, target_language: str, source_language: str = "auto"
    ) -> str:
        """
        Translate HTML content to target language.

        Args:
            content: HTML content to translate
            target_language: Target language code (e.g., 'en', 'de', 'es')
            source_language: Source language ('auto' to detect)

        Returns:
            Translated HTML content

        Raises:
            AIServiceError: On translation failure
        """
        system_prompt = """You are a professional translator. Translate the provided HTML content to the target language.

CRITICAL RULES:
1. Preserve ALL HTML tags, attributes, and structure exactly
2. Only translate text content within tags
3. Do NOT translate: URLs, code blocks, technical terms, proper nouns
4. Maintain formatting, line breaks, and spacing
5. Return ONLY the translated HTML in the 'translated_html' field"""

        user_prompt = f"""Translate this HTML content to {target_language}:

{content}"""

        messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ]

        # Structured output schema
        response_format = {
            "type": "json_schema",
            "json_schema": {
                "name": "translation_response",
                "strict": True,
                "schema": {
                    "type": "object",
                    "properties": {
                        "detected_language": {
                            "type": "string",
                            "description": "Detected source language code",
                        },
                        "translated_html": {
                            "type": "string",
                            "description": "Translated HTML content",
                        },
                    },
                    "required": ["detected_language", "translated_html"],
                    "additionalProperties": False,
                },
            },
        }

        result = self._make_request(messages, response_format)

        logger.info(
            f"Translated content from {result.get('detected_language', 'unknown')} to {target_language}"
        )

        return result["translated_html"]

    def summarize(self, content: str) -> str:
        """
        Generate concise summary of HTML content.

        Args:
            content: HTML content to summarize

        Returns:
            Summary as HTML (bullet points)

        Raises:
            AIServiceError: On summarization failure
        """
        system_prompt = """You are a content summarizer. Create a concise summary of the article.

RULES:
1. Extract 3-5 key points as bullet list
2. Each point should be 1-2 sentences max
3. Focus on main ideas, facts, and conclusions
4. Return HTML formatted list (<ul><li>...</li></ul>)
5. Be objective and factual"""

        user_prompt = f"""Summarize this article:

{content}"""

        messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ]

        response_format = {
            "type": "json_schema",
            "json_schema": {
                "name": "summary_response",
                "strict": True,
                "schema": {
                    "type": "object",
                    "properties": {
                        "summary_html": {
                            "type": "string",
                            "description": "Summary as HTML bullet list",
                        }
                    },
                    "required": ["summary_html"],
                    "additionalProperties": False,
                },
            },
        }

        result = self._make_request(messages, response_format)

        logger.info("Generated summary")

        return result["summary_html"]

    def custom_prompt(self, content: str, prompt: str) -> str:
        """
        Process content with custom user prompt.

        Args:
            content: HTML content to process
            prompt: User's custom instruction

        Returns:
            Processed content as HTML

        Raises:
            AIServiceError: On processing failure
        """
        system_prompt = f"""You are a content processor. Follow the user's instruction to process the article.

RULES:
1. Follow the user's instruction exactly
2. Return result as clean HTML
3. Preserve important formatting
4. Be concise and relevant

User Instruction: {prompt}"""

        user_prompt = f"""Process this article:

{content}"""

        messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ]

        response_format = {
            "type": "json_schema",
            "json_schema": {
                "name": "custom_prompt_response",
                "strict": True,
                "schema": {
                    "type": "object",
                    "properties": {
                        "processed_html": {
                            "type": "string",
                            "description": "Processed HTML content",
                        }
                    },
                    "required": ["processed_html"],
                    "additionalProperties": False,
                },
            },
        }

        result = self._make_request(messages, response_format)

        logger.info(f"Processed content with custom prompt: {prompt[:50]}...")

        return result["processed_html"]

    def check_quota(self, user) -> None:
        """
        Check if user has remaining AI quota.

        Args:
            user: Django User instance

        Raises:
            AIQuotaExceededError: If quota exceeded
        """
        from core.models import UserAIQuota

        # Get or create quota for user
        quota, created = UserAIQuota.objects.get_or_create(
            user=user,
            defaults={
                "daily_limit": settings.AI_DEFAULT_DAILY_LIMIT,
                "monthly_limit": settings.AI_DEFAULT_MONTHLY_LIMIT,
                "daily_reset_at": (timezone.now() + timedelta(days=1)).replace(
                    hour=0, minute=0, second=0, microsecond=0
                ),
                "monthly_reset_at": (
                    (timezone.now() + timedelta(days=32)).replace(
                        day=1, hour=0, minute=0, second=0, microsecond=0
                    )
                ),
            },
        )

        if not quota.can_use_ai():
            raise AIQuotaExceededError(
                f"AI quota exceeded for user {user.username}: "
                f"{quota.daily_used}/{quota.daily_limit} daily, "
                f"{quota.monthly_used}/{quota.monthly_limit} monthly"
            )

    def increment_quota(self, user) -> None:
        """
        Increment user's AI usage counter.

        Args:
            user: Django User instance
        """
        from core.models import UserAIQuota

        quota = UserAIQuota.objects.get(user=user)
        quota.increment_usage()

File: src/server/services/ai.service.ts
/**
 * AI service for content processing using OpenAI-compatible APIs.
 *
 * Provides translation, summarization, and custom prompt processing
 * with structured JSON output and retry logic.
 */

import axios from "axios";

import { logger } from "../utils/logger";

import { AIRequestRetryHandler } from "./ai-request-handler";
import { AIResponseParser } from "./ai-response-parser";
import type { AIServiceConfig } from "./ai.service.interface";

export class AIService {
  private config: AIServiceConfig;

  constructor(config: AIServiceConfig) {
    this.config = config;
  }

  /**
   * Replace base64 images with placeholders to reduce token usage.
   */
  private replaceBase64ImagesWithPlaceholders(
    content: string,
  ): [string, Map<string, string>] {
    const pattern =
      /(<img[^>]*\ssrc=)(["']?)(data:image\/[^;]+;base64,[^"'>\s]+)(\2)/gi;

    const placeholders = new Map<string, string>();
    let placeholderCounter = 0;

    const contentWithPlaceholders = content.replace(
      pattern,
      (match, prefix, quoteStart, base64Data, quoteEnd) => {
        const placeholder = `[IMAGE_PLACEHOLDER_${placeholderCounter}]`;
        placeholders.set(placeholder, base64Data);
        placeholderCounter++;
        return `${prefix}${quoteStart}${placeholder}${quoteEnd}`;
      },
    );

    if (placeholders.size > 0) {
      logger.debug(
        { count: placeholders.size },
        "Replaced base64 images with placeholders",
      );
    }

    return [contentWithPlaceholders, placeholders];
  }

  /**
   * Restore base64 images from placeholders.
   */
  private restoreBase64ImagesFromPlaceholders(
    content: string,
    placeholders: Map<string, string>,
  ): string {
    if (placeholders.size === 0) return content;

    let restored = content;
    for (const [placeholder, base64Data] of placeholders.entries()) {
      restored = restored.replace(placeholder, base64Data);
    }

    logger.debug(
      { count: placeholders.size },
      "Restored base64 images from placeholders",
    );
    return restored;
  }


  /**
   * Make API request with retry logic.
   */
  private async makeRequest(
    messages: Array<{ role: string; content: string }>,
    responseFormat?: {
      type: string;
      json_schema: {
        name: string;
        strict: boolean;
        schema: Record<string, unknown>;
      };
    },
  ): Promise<Record<string, unknown>> {
    const retryHandler = new AIRequestRetryHandler({
      maxRetries: this.config.maxRetries,
      retryDelay: this.config.retryDelay,
    });
    const responseParser = new AIResponseParser();

    const payload = this.buildRequestPayload(messages, responseFormat);
    const headers = this.buildRequestHeaders();

    let lastError: Error | null = null;

    for (let attempt = 0; attempt < this.config.maxRetries; attempt++) {
      try {
        const data = await this.executeRequest(payload, headers);
        return this.handleSuccessResponse(data, responseFormat, responseParser);
      } catch (error) {
        lastError = await this.handleRequestError(
          error,
          attempt,
          retryHandler,
          responseParser,
        );
      }
    }

    throw new Error(
      `AI request failed after ${this.config.maxRetries} retries: ${lastError?.message}`,
    );
  }

  /**
   * Build request payload.
   */
  private buildRequestPayload(
    messages: Array<{ role: string; content: string }>,
    responseFormat?: Record<string, unknown>,
  ): Record<string, unknown> {
    const payload: Record<string, unknown> = {
      model: this.config.model,
      messages,
      temperature: this.config.temperature,
      max_tokens: this.config.maxTokens,
    };

    if (responseFormat) {
      payload["response_format"] = responseFormat;
    }

    return payload;
  }

  /**
   * Build request headers.
   */
  private buildRequestHeaders(): Record<string, string> {
    return {
      Authorization: `Bearer ${this.config.apiKey}`,
      "Content-Type": "application/json",
    };
  }

  /**
   * Execute API request.
   */
  private async executeRequest(
    payload: Record<string, unknown>,
    headers: Record<string, string>,
  ): Promise<Record<string, unknown>> {
    const response = await axios.post(
      `${this.config.apiUrl}/chat/completions`,
      payload,
      {
        headers,
        timeout: this.config.timeout * 1000,
      },
    );
    return response.data;
  }

  /**
   * Handle successful API response.
   */
  private handleSuccessResponse(
    data: Record<string, unknown>,
    responseFormat: Record<string, unknown> | undefined,
    responseParser: AIResponseParser,
  ): Record<string, unknown> {
    const choices = data.choices as Array<Record<string, unknown>>;
    const message = choices[0].message as Record<string, unknown>;
    const contentStr = String(message.content);
    const finishReason = String(choices[0].finish_reason || "");

    if (responseParser.isTruncated(finishReason)) {
      responseParser.logTruncation(contentStr.length, this.config.maxTokens);
    }

    if (responseFormat) {
      return responseParser.parseJSON(
        contentStr,
        finishReason,
        this.config.maxTokens,
        this.config.maxRetries,
        0,
      );
    }

    return { content: contentStr };
  }

  /**
   * Handle request error with retry logic.
   */
  private async handleRequestError(
    error: unknown,
    attempt: number,
    retryHandler: AIRequestRetryHandler,
    responseParser: AIResponseParser,
  ): Promise<Error> {
    if (responseParser.isJsonParseError(error)) {
      const parsedError = error instanceof Error ? error : new Error(String(error));
      if (retryHandler.shouldRetry(attempt)) {
        await this.waitBeforeRetry(attempt, retryHandler, null);
      }
      return parsedError;
    }

    const retryInfo = retryHandler.extractRetryInfo(error);
    retryHandler.logRetryAttempt(
      attempt,
      retryInfo.isRateLimit,
      retryInfo.error.message,
    );

    if (retryHandler.shouldRetry(attempt)) {
      await this.waitBeforeRetry(attempt, retryHandler, retryInfo.retryAfter);
    }

    return retryInfo.error;
  }

  /**
   * Wait before retrying with appropriate delay.
   */
  private async waitBeforeRetry(
    attempt: number,
    retryHandler: AIRequestRetryHandler,
    retryAfter: number | null,
  ): Promise<void> {
    const delay = retryHandler.calculateRetryDelay(attempt, retryAfter);
    logger.info({ delay }, "Retrying after delay");
    await retryHandler.wait(delay);
  }

  /**
   * Translate HTML content to target language.
   */
  async translate(
    content: string,
    targetLanguage: string,
    _sourceLanguage: string = "auto",
  ): Promise<string> {
    const systemPrompt = `You are a professional translator. Translate the provided HTML content to the target language.

CRITICAL RULES:
1. Preserve ALL HTML tags, attributes, and structure exactly
2. Only translate text content within tags
3. Do NOT translate: URLs, code blocks, technical terms, proper nouns
4. Maintain formatting, line breaks, and spacing
5. Preserve image placeholders (e.g., [IMAGE_PLACEHOLDER_0]) exactly as they appear
6. Return ONLY the translated HTML in the 'translated_html' field`;

    // Replace base64 images with placeholders
    const [contentWithPlaceholders, placeholders] =
      this.replaceBase64ImagesWithPlaceholders(content);

    const userPrompt = `Translate this HTML content to ${targetLanguage}:

${contentWithPlaceholders}`;

    const messages = [
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt },
    ];

    const responseFormat = {
      type: "json_schema",
      json_schema: {
        name: "translation_response",
        strict: true,
        schema: {
          type: "object",
          properties: {
            detected_language: {
              type: "string",
              description: "Detected source language code",
            },
            translated_html: {
              type: "string",
              description: "Translated HTML content",
            },
          },
          required: ["detected_language", "translated_html"],
          additionalProperties: false,
        },
      },
    };

    const result = (await this.makeRequest(
      messages,
      responseFormat,
    )) as unknown as {
      detected_language: string;
      translated_html: string;
    };

    // Restore base64 images
    const translatedHtml = this.restoreBase64ImagesFromPlaceholders(
      result.translated_html,
      placeholders,
    );

    logger.info(
      {
        detectedLanguage: result.detected_language,
        targetLanguage,
      },
      "Content translated",
    );

    return translatedHtml;
  }

  /**
   * Generate concise summary of HTML content.
   */
  async summarize(content: string): Promise<string> {
    const systemPrompt = `You are a content summarizer. Create a concise summary of the article.

RULES:
1. Extract 3-5 key points as bullet list
2. Each point should be 1-2 sentences max
3. Focus on main ideas, facts, and conclusions
4. Return HTML formatted list (<ul><li>...</li></ul>)
5. Be objective and factual
6. Keep the summary in the content language (do not translate it yet)
7. Preserve image placeholders (e.g., [IMAGE_PLACEHOLDER_0]) if present in summary`;

    // Replace base64 images with placeholders
    const [contentWithPlaceholders, placeholders] =
      this.replaceBase64ImagesWithPlaceholders(content);

    const userPrompt = `Summarize this article:

${contentWithPlaceholders}`;

    const messages = [
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt },
    ];

    const responseFormat = {
      type: "json_schema",
      json_schema: {
        name: "summary_response",
        strict: true,
        schema: {
          type: "object",
          properties: {
            summary_html: {
              type: "string",
              description: "Summary as HTML bullet list",
            },
          },
          required: ["summary_html"],
          additionalProperties: false,
        },
      },
    };

    const result = (await this.makeRequest(
      messages,
      responseFormat,
    )) as unknown as {
      summary_html: string;
    };

    // Restore base64 images
    const summaryHtml = this.restoreBase64ImagesFromPlaceholders(
      result.summary_html,
      placeholders,
    );

    logger.info("Content summarized");
    return summaryHtml;
  }

  /**
   * Process content with custom prompt.
   */
  async processCustomPrompt(content: string, prompt: string): Promise<string> {
    const systemPrompt = `You are a content processor. Process the provided HTML content according to the user's instructions.

CRITICAL RULES:
1. Preserve ALL HTML tags, attributes, and structure exactly
2. Only modify text content as instructed
3. Do NOT translate unless explicitly asked
4. Maintain formatting, line breaks, and spacing
5. Preserve image placeholders (e.g., [IMAGE_PLACEHOLDER_0]) exactly as they appear
6. Return ONLY the processed HTML in the 'processed_html' field`;

    // Replace base64 images with placeholders
    const [contentWithPlaceholders, placeholders] =
      this.replaceBase64ImagesWithPlaceholders(content);

    const userPrompt = `${prompt}

Content to process:

${contentWithPlaceholders}`;

    const messages = [
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt },
    ];

    const responseFormat = {
      type: "json_schema",
      json_schema: {
        name: "custom_prompt_response",
        strict: true,
        schema: {
          type: "object",
          properties: {
            processed_html: {
              type: "string",
              description: "Processed HTML content",
            },
          },
          required: ["processed_html"],
          additionalProperties: false,
        },
      },
    };

    const result = (await this.makeRequest(
      messages,
      responseFormat,
    )) as unknown as {
      processed_html: string;
    };

    // Restore base64 images
    const processedHtml = this.restoreBase64ImagesFromPlaceholders(
      result.processed_html,
      placeholders,
    );

    logger.info("Content processed with custom prompt");
    return processedHtml;
  }
}

/**
 * Create AI service instance from user settings.
 */
export function createAIService(userSettings: {
  openaiApiUrl: string;
  openaiApiKey: string;
  aiModel: string;
  aiTemperature: number;
  aiMaxTokens: number;
  aiRequestTimeout: number;
  aiMaxRetries: number;
  aiRetryDelay: number;
}): AIService {
  if (!userSettings.openaiApiKey) {
    throw new Error("OpenAI API key not configured");
  }

  return new AIService({
    apiUrl: userSettings.openaiApiUrl.replace(/\/$/, ""),
    apiKey: userSettings.openaiApiKey,
    model: userSettings.aiModel,
    temperature: userSettings.aiTemperature,
    maxTokens: userSettings.aiMaxTokens,
    timeout: userSettings.aiRequestTimeout,
    maxRetries: userSettings.aiMaxRetries,
    retryDelay: userSettings.aiRetryDelay,
  });
}
